### PR TITLE
Simple forking and merging feature

### PR DESCRIPTION
**Pull request by viktor992**
**Changes:**
- Fork and Merge buttons added to Sidebar.
- `Document.fork` and `Document.merge` methods added.
- Undo changes button added for debugging purposes.
- Added migration for adding `forkedAtVersion` and `forkedFrom` fields to Document collection.

**Steps to forking/merging documents:** 
- Create a new document and add some content.
- Click Fork button. A new document with the same content will be created and the user will be redirected to the forked document editor.
- Edit the forked and original documents.
- Click Merge button on the forked document. The steps will be applied to the original document. The user will be redirected to original document editor.
- The undo changes button works only on forked (non merged) documents for now.

**Pending changes/fixes:**
- Fix migration for changing `Content.contentKey` to `Content.contentKeys`.

See #139

### CODE REVIEW

File: client/components/app-layout.vue [hunk 1]
```diff
@@ -56,7 +56,7 @@
         fluid
         fill-height
       >
-        <router-view />
+        <router-view :key="$route.fullPath" />
```
Comment (mitar) [inline client/components/app-layout.vue]: Why is this necessary?

File: client/components/document.vue [hunk 2]
```diff
@@ -72,6 +75,10 @@
     },
 
     methods: {
+      reloadEditor(key) {
```
Comment (mitar) [inline client/components/document.vue]: So this is just temporary?

File: client/components/editor.vue [hunk 3]
```diff
@@ -115,7 +115,6 @@
         </v-btn-toggle>
 
         <v-spacer />
-
```
Comment (mitar) [inline client/components/editor.vue]: No need to change this.

File: lib/documents/document.js [hunk 4]
```diff
@@ -28,6 +28,8 @@ export class Document extends BaseDocument {
   // contentKey: ID used to identify editor's content
```
Comment (mitar) [inline lib/documents/document.js]: You have to rename this.

File: lib/documents/document.js [hunk 5]
```diff
@@ -36,6 +38,8 @@ export class Document extends BaseDocument {
       publishedAt: 1,
       title: 1,
       contentKey: 1,
+      forkedFrom: 1,
+      forkedAtVersion: 1,
```
Comment (mitar) [inline lib/documents/document.js]: Do you have to publish these fields to the client?

File: server/documents/content.js [hunk 6]
```diff
@@ -28,7 +32,11 @@ Content.getCurrentState = (args) => {
   }
 
   Content.documents.find({
-    contentKey: args.contentKey,
+    contentKeys: {
```
Comment (mitar) [inline server/documents/content.js]: Mongo allows just `contentKeys: args.contentKey`.

File: server/documents/content.js [hunk 7]
```diff
@@ -217,11 +234,22 @@ Meteor.publish('Content.list', function contentList(args) {
...
     }
 
-    return Content.documents.find({
-      contentKey: args.contentKey,
-    }, {
-      fields: Content.PUBLISH_FIELDS(),
-    });
+    return Content.documents.find(
+      {
+        $or: [
```
Comment (mitar) [inline server/documents/content.js]: Why `$or` here?

File: server/documents/document.js [hunk 8]
```diff
@@ -44,6 +48,291 @@ Meteor.methods({
...
+    const {contentKey} = doc;
+
+    const createdAt = new Date();
+    const forkContentKey = Content.Meta.collection._makeNewID();
+
+    // Add forkContentKey to the existing contents.
+    Content.documents.update(
+      {
+        contentKeys:
+        {
```
Comment (mitar) [inline server/documents/document.js]: This `{` can be moved one line up, no?

File: server/documents/document.js [hunk 9]
```diff
@@ -44,6 +48,291 @@ Meteor.methods({
...
+          $elemMatch: {
```
Comment (mitar) [inline server/documents/document.js]: I do not think you need `$elemMatch` here. You are matching only against one field.

File: server/documents/document.js [hunk 10]
```diff
@@ -44,6 +48,291 @@ Meteor.methods({
...
+            $in: [contentKey],
+          },
+        },
+      },
+      {
+        $push:
```
Comment (mitar) [inline server/documents/document.js]: Maybe use `$addToSet` just to be sure.

File: server/documents/document.js [hunk 11]
```diff
@@ -44,6 +48,291 @@ Meteor.methods({
...
+
+    // Save merge steps.
+    transform.steps.forEach((x, i) => {
+      if (i >= originalSteps.length) {
+        version += 1;
+        Content.documents.upsert({
+          version,
+          contentKeys: {
+            $elemMatch: {
+              $in: [original.contentKey],
```
Comment (mitar) [inline server/documents/document.js]: Hmm. Are you sure this is correct?

Where do you remove steps done on the original document after the fork point? Or do you add both revert steps and then rebased steps?

Why are you using here upsert if you are not really updating anything?

**Comment by mitar** [comment]
So one thing we have to make sure that it works well if the original document is forked twice:
 * In that case after merging the first fork, the original document has to be rebased.
 * But also the second fork has to be rebased so that now it is forked from now new rebased original document. I will have to think more how things should happen.

We should store somewhere information which steps came from which fork, and where/when were merges. So that in the future when displaying history we can show that. Not sure if that should be `Content` or `Document` documents. Probably the latter?

One tricky thing I also think will be is how to do all those changes to MongoDB without having conflicts at the same.

**Comment by mitar** [comment]
One more thing, when forking, comments should not be forked, but highlights should. (I think this is already so, just noting.)

File: client/components/app-layout.vue (hunk 1 above)
Comment (viktor992) [reply client/components/app-layout.vue]: Suppose that an user is editing a document (*d1*) at the */document/d1* url and clicks on the Fork button. A new document (*d2*) is created and the user has to be redirected to */document/d2* (basically to the same page but displaying the new document). If `:key="$route.fullPath"` is not used in the `router-view` the document component won't be reloaded because we are routing to the same component.

File: client/components/document.vue (hunk 2 above)
Comment (viktor992) [reply client/components/document.vue]: I added this just for reloading the `editor` component after the *Undo Changes* button is pressed. We need to reload the prosemirror editor state (basically rerun what's inside of `mounted`).

File: lib/documents/document.js (hunk 4 above)
Comment (viktor992) [reply lib/documents/document.js]: Do you mean that I have to rename the `contentKey` field or that I have to change the field description?

File: lib/documents/document.js (hunk 5 above)
Comment (viktor992) [reply lib/documents/document.js]: I use these fields to toggle the Fork/Merge buttons in the sidebar component. Should I add a boolean field for this? something like `isForked` maybe?

File: server/documents/document.js (hunk 11 above)
Comment (viktor992) [reply server/documents/document.js]: Yes I do, I add both revert steps and then rebased steps. I'll use insert instead of upsert.

File: server/documents/content.js (hunk 7 above)
Comment (viktor992) [reply server/documents/content.js]: Unnecessary.

File: server/documents/content.js [hunk 12]
```diff
@@ -27,6 +34,252 @@ function updateCurrentState(contentKey, doc, version) {
   }
 }
 
+function rebaseSteps(args) {
+  const {documentId} = args;
+
+  // TODO: Check Rebase permissions.
+
+  // Get forked document.
+  const fork = Document.documents.findOne(
```
Comment (mitar) [inline server/documents/content.js]: This rebases only the first forked document, no?

**Comment by mitar** [comment]
I made some changes, mostly adding more locking.

**Comment by mitar** [comment]
Also, added that rebasing is recursively done.
